### DISCUSSION TITLE
EntityView with fields referencing items of a mapped collection

### DISCUSSION THREAD

**Discussion by cepxy** [Blazebit/blaze-persistence#1430] [Q&A]
Is it possible to build an entity view with fields that reference dedicated/filtered entries of a mapped collection?

Say we have an entity with some simple fields and a list of status updates. Over time that list gets filled with more and more status updates.
We now want an overview of the entities with a subset of fields and dedicated fields for some specific status updates that we are interested in.
If an entities status list contains these status updates already, the respective field will be set, otherwise it is null.

See this example and the EntityView we would like to have:

```
class Example {
Id id;
String name;
String foo;
String bar;
List<StatusUpdate> statusUpdates; // (could contain status "a" to "z")
}

class StatusUpdate {
LocalDateTime date;
Status status;
}

@EntityView(Example.class)
interface ExampleView {
String getName();
// add some mapping and filtering
StatusUpdate getStatusA(); // contain status "a" if existing in list statusUpdates
// add some mapping and filtering
StatusUpdate getStatusX(); // contain status "x" if existing in list statusUpdates
}
```

Is that somehow possible with blaze persistance to setup fields that reference items of a collection (if existing) based in specified criteria?
We played around with filtering, but did not get it to work to map specific status updates to single fields.

What we did get to work is a ViewFilterProvider where we select all the status updates we are interested in and map it to a whole list of Status updates.
So we end up with one list again, but at least it contains only the desired status updates which we can process manually later.

```
@EntityView(Example.class)
@ViewFilter(name = "statusFilter", value = ExampleView.StatusFilterProvider.class)
interface ExampleView {
String getName();
List<StatusUpdate> getStatusUpdates(); // filtered list of status updates "a" and "x"
}

class StatusFilterProvider extends ViewFilterProvider {
	@Override
	public <T extends WhereBuilder<T>> T apply(T whereBuilder) {
		return whereBuilder
				.whereOr()
				.where("statusUpdates.status.key").eq("a")
				.where("statusUpdates.status.key").eq("x")
				.endOr();
	}
}
```
and applying: `setting.addViewFilter("statusFilter");`

However that has (at least) one issue: Entities that dont have any of the desired status updates at all will be left out of the result. Not good.
As a workaround we added a (initial) status that all entities have anyway to the filter, although we wont need it later.

Are there any other ways to do it?

Thanks and regards,
Christian

**Comment by beikov** [comment]
Yeah, it's pretty simple to do that with array expressions like this:

```java
@EntityView(Example.class)
interface ExampleView {

    String getName();

    @Mapping("statusUpdates[status = fully.qualified.name.Status.STATUS_A]")
    StatusUpdate getStatusA();
    
    @Mapping("statusUpdates[status = fully.qualified.name.Status.STATUS_X]")
    StatusUpdate getStatusX();
}
```

**Comment by cepxy** [comment]
Thanks for the quick reply, that helped a lot. Got a bit further.
Following worked:
```
@Mapping("statusUpdates[status.key = 'status_a']")
StatusUpdate getStatusA();
```
But I couldn't use a direct reference to the actual enum-reference. Although our Model is a bit more complex, as we are not using simple enums but actual database entries reflecting the different status. There is a enum with references to the matching key, but I cant use this in the expression. So follwing did not work, even if I used the full qualifier:
```
@Mapping("statusUpdates[status.key = StatusRef.STATUS_A.key]")
StatusUpdate getStatusA();
```
```
java.lang.IllegalArgumentException: There are error(s) in entity views!
Exception while resolving type for attribute submitted[...ExampleView.getStatusA]:
java.lang.IllegalArgumentException: Attribute 'StatusRef' not found on type 'StatusUpdate'
```

One more question:
In our realworld example, status updates with the same status can happen multiple times. That means for our view we have to use lists (or otherwise end up with duplicate entries of the main entity).
We are only interested in the last (or first) status update of a certain status. So the question is: can the mapping be extended to have a criteria to order and return only one entry? Something similar to 'TOP 1'?

**Reply by beikov** [reply]
The fully qualified name approach only works for real enums. You can't de-reference the enum unfortunately.

What you are looking for is the `@Limit` mapping. 

```java
@EntityView(Example.class)
interface ExampleView {

    String getName();

    @Limit(limit = "1", order = "id desc")
    @Mapping("statusUpdates[status = fully.qualified.name.Status.STATUS_A]")
    StatusUpdate getStatusA();
    
    @Limit(limit = "1", order = "id desc")
    @Mapping("statusUpdates[status = fully.qualified.name.Status.STATUS_X]")
    StatusUpdate getStatusX();
}
```

Here is the documentation for more details: https://persistence.blazebit.com/documentation/1.6/entity-view/manual/en_US/#limit-mapping
